corehq/apps/hqcase/api/get_list.py

```diff
@@ -80,35 +87,54 @@ def get_list(domain, params):
 
     cases_in_result = len(hits)
     if cases_in_result and es_result.total > cases_in_result:
-        cursor = urlencode({**params, **{'indexed_on.gte': hits[-1]["@indexed_on"]}})
+        cursor = urlencode({**params, **{
+            INDEXED_AFTER: hits[-1]["@indexed_on"],
+            LAST_CASE_ID: hits[-1]["_id"],
+        }})
         ret['next'] = {'cursor': b64encode(cursor.encode('utf-8'))}
 
     return ret
 
 
-def _run_query(domain, params):
-    params = params.copy()
-    page_size = _to_int(params.pop('limit', DEFAULT_PAGE_SIZE), 'limit')
+def _get_cursor_query(domain, params, last_date, last_id):
+    query = _get_query(domain, params)
+    return query.filter(
+        filters.OR(
+            filters.AND(
+                filters.term('@indexed_on', last_date),
+                filters.range_filter('_id', gt=last_id),
+            ),
+            case_search.indexed_on(gt=last_date),
+        )
+    )
```

Comment:
@snopoke I'm interested to hear your take on this - it skips the last seen ID in the query rather than in memory.

I was reviewing your implementation of cursor pagination in the messaging event API, and it occurred to me that if there are ever more than two rows with identical sort dates, there's still a possibility for overlap between pages, whereas I think this handles it.

Reply:
:+1: this is the approach we take in the re-index accessors: https://github.com/dimagi/commcare-hq/blob/cda566bbce803bb03edb71bb7436caa6bbcdcca1/corehq/form_processor/backends/sql/dbaccessors.py#L286

My only concern here is that the last time I tried sorting in this index it had very poor performance.

Reply:
Yeah, the sort performance on prod is the biggest blocker to releasing this API - it's unusable at our current scale.  The hope is that the next ES upgrade will fix it.  I hadn't considered that adding `_id` as a secondary sort might worsen the performance, though.  If this isn't resolved with the upgrade, we might need to look into something like [this](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html#search-after)